Userland/Utilities/uniq.cpp

```diff
@@ -5,75 +5,122 @@
+    StringView inpath;
+    StringView outpath;
+    bool duplicates_only = false;
+    bool unique_only = false;
+    bool ignore_case = false;
+    bool print_count = false;
+    unsigned int skip_chars = 0;
+    unsigned int skip_fields = 0;
+
     Core::ArgsParser args_parser;
+    args_parser.add_option(duplicates_only, "Only print duplicated lines", "repeated", 'd');
+    args_parser.add_option(unique_only, "Only print unique lines (default)", "unique", 'u');
+    args_parser.add_option(ignore_case, "Ignore case when comparing lines", "ignore-case", 'i');
+    args_parser.add_option(print_count, "Prefix each line by its number of occurences", "count", 'c');
+    args_parser.add_option(skip_chars, "Skip N chars", "skip-chars", 's', "N");
+    args_parser.add_option(skip_fields, "Skip N fields", "skip-fields", 'f', "N");
     args_parser.add_positional_argument(inpath, "Input file", "input", Core::ArgsParser::Required::No);
     args_parser.add_positional_argument(outpath, "Output file", "output", Core::ArgsParser::Required::No);
     args_parser.parse(arguments);
 
-    FILE* infile = get_stream(inpath, "r");
-    FILE* outfile = get_stream(outpath, "w");
-
-    struct linebuf buffers[2];
-    struct linebuf* previous = &(buffers[0]);
-    struct linebuf* current = &(buffers[1]);
-    bool first_run = true;
-    for (;;) {
-        errno = 0;
-        ssize_t rc = getline(&(current->buf), &(current->len), infile);
-        if (rc < 0 && errno != 0) {
-            perror("getline");
-            exit(1);
-        }
-        if (rc < 0)
-            break;
-        if (!first_run && strcmp(current->buf, previous->buf) == 0)
-            continue;
+    if (!unique_only && !duplicates_only)
+        unique_only = true;
+    else if (unique_only && duplicates_only) {
+        // Printing duplicated and unique lines shouldn't print anything
+        return 0;
+    }
+
+    if (outpath.is_empty() || outpath == "-"sv)
+        outpath = "/dev/stdout"sv;
+
+    auto infile = TRY(open_file_or_stdin(inpath));
+    auto outfile = TRY(Core::Stream::File::open(outpath, Core::Stream::OpenMode::Write));
+
+    size_t count = 0;
+    ByteBuffer previous_buf = TRY(ByteBuffer::create_uninitialized(1024));
+    ByteBuffer current_buf = TRY(ByteBuffer::create_uninitialized(1024));
 
-        fputs(current->buf, outfile);
+    StringView previous = TRY(infile->read_line(previous_buf));
+    StringView previous_to_compare = skip(previous, skip_chars, skip_fields);
+
+    while (TRY(infile->can_read_line())) {
+        StringView current = TRY(infile->read_line(current_buf));
+
+        StringView current_to_compare = skip(current, skip_chars, skip_fields);
+        bool lines_equal = ignore_case ? current_to_compare.equals_ignoring_case(previous_to_compare) : current_to_compare == previous_to_compare;
+        if (!lines_equal) {
+            TRY(write_line_content(previous, count, duplicates_only, print_count, *outfile));
+            count = 0;
```

Comment:
Maybe we should make the counter here one-based instead of zero-based. Seeing a new string and then going "this is the zeroeth time that I have seen that line" is not immediately obvious.